### PR TITLE
Make node serial cross platform, thanks to nmcspadden for windows method

### PR DESCRIPTION
**Pull request by natewalck**
This makes `node.serial` cross platform for macOS, Windows and Debian based OSes. We don't have anything redhat/fedora, but I'd love to see an implementation there too :) 

Thanks to @nmcspadden for the windows method here. It offers better support across multiple machine types.

### CODE REVIEW

**Comment by johnnyramos** [comment]
Don't forget to review the CI and decide whether it's worth fixing or Facebook needs to tweak rules:
```
Offenses:
chef/cookbooks/cpe_utils/libraries/node_functions.rb:47:7: C: Layout/EmptyLineAfterGuardClause: Add empty line after guard clause.
      return default if path.nil?
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
chef/cookbooks/cpe_utils/libraries/node_functions.rb:143:7: C: Layout/EmptyLineAfterGuardClause: Add empty line after guard clause.
      return node['uuid'] if node['uuid']
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
chef/cookbooks/cpe_utils/libraries/node_functions.rb:212:7: C: Layout/EmptyLineAfterGuardClause: Add empty line after guard clause.
      return false unless st
      ^^^^^^^^^^^^^^^^^^^^^^
chef/cookbooks/cpe_utils/libraries/node_functions.rb:482:1: C: Layout/EmptyLines: Extra blank line detected.
chef/cookbooks/cpe_utils/libraries/node_functions.rb:484:5: C: Layout/EmptyLineBetweenDefs: Use empty lines between method definitions.
    def _parse_profiles_status(output)
    ^^^
chef/cookbooks/cpe_utils/libraries/node_functions.rb:501:7: C: Layout/EmptyLineAfterGuardClause: Add empty line after guard clause.
      fail 'profiles XML parsing must be a Hash!' unless profiles.is_a?(Hash)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
chef/cookbooks/cpe_utils/libraries/node_functions.rb:564:7: C: Style/UnneededCondition: This condition is not needed.
      node['virtual_macos'] if node['virtual_macos']
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by natewalck** [comment]
I didn't change any of those lines :)

**Comment by johnnyramos** [comment]
My goodness, and that's how contributions die so quick - "`it ain't my fault`".

**Comment by natewalck** [comment]
Hahaha. It was mild trolling :) 
    	
    	

    	
    
  




On Wed, Mar 20, 2019 at 10:50 PM -0400, "Johnny Ramos" <notifications@github.com> wrote:












My goodness, and that's how contributions die so quick - "it ain't my fault".

**Comment by johnnyramos** [comment]
Helping you out a bit: https://github.com/natewalck/IT-CPE/pull/1

**Comment by natewalck** [comment]
Thanks @johnnyramos! I'll look at the rest then...

**Comment by johnnyramos** [comment]
No worries, I don't mind. Anotha one: https://github.com/natewalck/IT-CPE/pull/2

**Comment by johnnyramos** [comment]
Ping. :)

**Comment by natewalck** [comment]
I'm happy with this. Waiting on FB for the merge :)

File: chef/cookbooks/cpe_utils/libraries/node_functions.rb [hunk 1]
```diff
@@ -100,17 +101,39 @@ def console_user
     end
 
     def serial
```
Comment (nmcspadden) [inline chef/cookbooks/cpe_utils/libraries/node_functions.rb]:
```suggestion
        def serial_number
      @system_serial_number ||= begin
        system_serial = nil

        if linux?
          # TODO: T32760583 Stop using system-uuid here
          # Should probably stop using anything but system-serial-number
          # Try really hard to set the serial number.
          # Please keep these in this order unless you have a good reason to
          # change the ordering. Regardless, keep system-uuid last, it's the
          # last resort option.
          %w{
            system-serial-number
            chassis-serial-number
            baseboard-serial-number
            system-uuid
          }.each do |dmistr|
            serial = shell_out(
              "/usr/sbin/dmidecode -s #{dmistr}",
            ).stdout.lines.map(&:chomp)[0]
            unless serial.nil? || /\s/.match(serial) || /^0?$/.match(serial)
              system_serial = serial
              break
            end
          end

          # Make sure the serial has no dashes
          system_serial.delete!('-')

          # Serials over 15 chars cause issues with AD binding
          if system_serial.length > 15
            system_serial = system_serial[0..14]
          end
        elsif macos?
          ioreg_stdout = shell_out(
            '/usr/sbin/ioreg -k IOPlatformSerialNumber -d 2 -a',
          ).stdout
          doc = Nokogiri::XML(ioreg_stdout)
          system_serial = doc.at_xpath(
            '/plist/dict//key[text()="IOPlatformSerialNumber"]/' +
            'following-sibling::string/text()',
          ).to_s
        elsif windows?
          require 'chef/mixin/powershell_out'
          serial_cmd = <<-'EOF'
Get-CimInstance Win32_ComputerSystemProduct | Select-Object IdentifyingNumber | ConvertTo-Json
EOF
          id_json = powershell_out!(serial_cmd).stdout
          begin
            serial_json = JSON.parse(id_json)
            system_serial = serial_json['IdentifyingNumber']
          rescue JSON::ParserError => e
            Chef::Log.error(
              "node/uuid_serial: Error: #{e.message}",
            )
            system_serial = nil
          end
        end

        system_serial
      end
    end
```

File: chef/cookbooks/cpe_utils/libraries/node_functions.rb [hunk 2]
```diff
@@ -100,17 +101,39 @@ def console_user
...
-        return
-      end
-      return node['serial'] if node['serial']
-      serial = Mixlib::ShellOut.new(
-        '/usr/sbin/ioreg -c IOPlatformExpertDevice |head -30' +
-        '|grep IOPlatformSerialNumber | awk \'{print $4}\' | sed -e s/\"//g',
-      ).run_command.stdout.chomp
-      node.default['serial'] = serial
-      serial
+      @serial ||=
```
Comment (nmcspadden) [inline chef/cookbooks/cpe_utils/libraries/node_functions.rb]:
```suggestion
      @system_serial_number ||= begin
        system_serial = nil

        if linux?
          # TODO: T32760583 Stop using system-uuid here
          # Should probably stop using anything but system-serial-number
          # Try really hard to set the serial number.
          # Please keep these in this order unless you have a good reason to
          # change the ordering. Regardless, keep system-uuid last, it's the
          # last resort option.
          %w{
            system-serial-number
            chassis-serial-number
            baseboard-serial-number
            system-uuid
          }.each do |dmistr|
            serial = shell_out(
              "/usr/sbin/dmidecode -s #{dmistr}",
            ).stdout.lines.map(&:chomp)[0]
            unless serial.nil? || /\s/.match(serial) || /^0?$/.match(serial)
              system_serial = serial
              break
            end
          end

          # Make sure the serial has no dashes
          system_serial.delete!('-')

          # Serials over 15 chars cause issues with AD binding
          if system_serial.length > 15
            system_serial = system_serial[0..14]
          end
        elsif macos?
          ioreg_stdout = shell_out(
            '/usr/sbin/ioreg -k IOPlatformSerialNumber -d 2 -a',
          ).stdout
          doc = Nokogiri::XML(ioreg_stdout)
          system_serial = doc.at_xpath(
            '/plist/dict//key[text()="IOPlatformSerialNumber"]/' +
            'following-sibling::string/text()',
          ).to_s
        elsif windows?
          require 'chef/mixin/powershell_out'
          serial_cmd = <<-'EOF'
Get-CimInstance Win32_ComputerSystemProduct | Select-Object IdentifyingNumber | ConvertTo-Json
EOF
          id_json = powershell_out!(serial_cmd).stdout
          begin
            serial_json = JSON.parse(id_json)
            system_serial = serial_json['IdentifyingNumber']
          rescue JSON::ParserError => e
            Chef::Log.error(
              "node/uuid_serial: Error: #{e.message}",
            )
            system_serial = nil
          end
        end

        system_serial
      end
```

Comment (nmcspadden) [reply chef/cookbooks/cpe_utils/libraries/node_functions.rb]: Here's a suggestion to improve on this a bit with our internal copy:

**Comment by natewalck** [comment]
This still needs to be tested, but of note:

1. Serial should not be truncated in the node function because you happen to use it for binding to AD. I've removed this :) 
2. Rubocop says the memoized variable name should match the function name, so I made it thus.

**Comment by natewalck** [comment]
@johnnyramos I can't get it to stop complaining about the stupid EOF :) I tried alllll the spots on that line.

**Comment by 1dustindavis** [comment]
Is this just waiting on the lint warning? Or is there more needed before this could be merged?

**Comment by jessepeterson** [comment]
Push this again please because I think we fixed the linter. :)

**Comment by michel-slm** [comment]
This, we can't merge as it's touching an old cookbook. We now use fb_helpers' node.serial which uses upstream Chef's Ohai plugin.

**Pull request closed without merging**